src/server.ts

```diff
@@ -26,120 +33,170 @@ export class MyAgent extends Agent<Env> {
-      }),
-      execute: async ({ recipient, subject, contentType, body }) => {
-        console.log({ recipient, subject, contentType, body })
-
-        try {
-          const success = await sendEmail(this.ctx.id, this.env.EMAIL, 'glen@gmad.dev', recipient, subject, contentType, body)
-
-          this.messages.push({
-            role: 'system',
-            content: `You have successfully sent the email. Please inform the user, with a summary of what you sent.`,
-          })
-
-          this.#talkToLLM(connection, false).then(async () => {
-            await scheduler.wait(2_000)
-
-            this.messages.push({
-              role: 'system',
-              content: `You have received the reply: "FUCK YES"`,
-            })
-            await this.#talkToLLM(connection, false)
-          })
-
-          return success
-        } catch (e) {
-          return `Error: ${(e as any).message}`
-        }
-      },
+  #sendEmail = tool({
+    description: 'Send a text or HTML email to an arbitrary recipient. You must specify an exact email address, do not invent one.',
+    parameters: z.object({
+      recipient: z.string().describe(`The email address of the recipient.`),
+      subject: z.string().describe(`The subject of the email.`),
+      contentType: z.string().describe(`The content type of the email. Can be text/plain or text/html`),
+      body: z.string().describe(`The body of the email. Must match the provided contentType parameter`),
+    }),
+    execute: async ({ recipient, subject, contentType, body }) => {
+      console.log({ recipient, subject, contentType, body })
+
+      try {
+        const success = await sendEmail(
+          this.ctx.id,
+          this.env.EMAIL,
+          FROM.address,
+          FROM.name,
+          FROM.domain,
+          recipient,
+          subject,
+          contentType,
+          body
+        )
+
+        this.messages.push({
+          role: 'system',
+          content: `The email system replied with: ${JSON.stringify(success)}. Please inform the user, with a summary of what you sent.`,
+        })
+
+        // Queue up an LLM call but return immediately
+        this.#talkToLLM(false)
+        return success
+      } catch (e) {
+        return `Error: ${(e as any).message}`
+      }
+    },
+  })
+
+  async receiveEmail(from: string, to: string, subject: string, contents: string) {
+    console.log({ id: this.ctx.id, from, to, subject, contents })
+
+    this.messages.push({
+      role: 'system',
+      content: [
+        `We have received an email!`,
+        `Update the user if this is in response to some conversation they're involved in or information they're waiting for.`,
+        `Message follows:`,
+        `---`,
+        `From: ${JSON.stringify(from)}`,
+        `To: ${JSON.stringify(to)}`,
+        `Subject: ${JSON.stringify(subject)}`,
+        `---`,
+        contents,
+      ].join('\n'),
```

Comment:
This seemed to do a good job, although the output on the frontend looks like arse. Not sure who's fault that is, but I blame the robot.